odl/space/weighting.py

```diff
@@ -0,0 +1,1047 @@
+﻿# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""CPU implementations of ``n``-dimensional Cartesian spaces.
+
+This is a default implementation of :math:`A^n` for an arbitrary set
+:math:`A` as well as the real and complex spaces :math:`R^n` and
+:math:`C^n`. The data is represented by NumPy arrays.
+"""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+# External module imports
+from math import sqrt
+import numpy as np
+import scipy.linalg as linalg
+from scipy.sparse.base import isspmatrix
+
+# ODL imports
+from odl.space.base_ntuples import FnBaseVector
+from odl.util.utility import array1d_repr, arraynd_repr
+
+
+__all__ = ('MatrixWeighting', 'VectorWeighting', 'ConstWeighting',
+           'NoWeighting',
+           'CustomInnerProduct', 'CustomNorm', 'CustomDist')
+
+
+class WeightingBase(object):
+
+    """Abstract base class for weighting of finite-dimensional spaces.
+
+    This class and its subclasses serve as a simple means to evaluate
+    and compare weighted inner products, norms and metrics semantically
+    rather than by identity on a pure function level.
+
+    The functions are implemented similarly to `Operator`,
+    but without extra type checks of input parameters - this is done in
+    the callers of the `LinearSpace` instance where these
+    functions are being used.
+    """
+
+    def __init__(self, impl, exponent=2.0, dist_using_inner=False):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        impl : `str`
+            Specifier for the implementation backend
+        exponent : positive `float`, optional
+            Exponent of the norm. For values other than 2.0, the inner
+            product is not defined.
+        dist_using_inner : `bool`, optional
+            Calculate `dist` using the following formula::
+
+                ||x - y||^2 = ||x||^2 + ||y||^2 - 2 * Re <x, y>
```

Comment:
That's the double colon, right. I guess it's better not to make it look like code.